File: CSharp/ExtensibleCommands/ExtensibleCommands/CyclicCommand.cs
namespace ExtensibleCommands
{
    /// <summary>
    /// Implements command to be repeated a specified number of times
    /// </summary>
    public class CyclicCommand : DecoratorCommand
    {
        /// <summary> How many times to repeat the core command </summary>
        public int NumberOfRepeats { get; }

        /// <summary> Current cycle number </summary>
        public int CurrentCycle { get; private set; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="numberOfRepeats">How many times to repeat the core command</param>
        public CyclicCommand(ICommand coreCommand, int numberOfRepeats)
            : this(coreCommand, numberOfRepeats, "Cyclic") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="numberOfRepeats">How many times to repeat the core command</param>
        /// <param name="name">Command name</param>
        public CyclicCommand(ICommand coreCommand, int numberOfRepeats, string name)
            : base(coreCommand, name)
        {
            NumberOfRepeats = numberOfRepeats;
        }

        protected override void Execute()
        {
            CurrentCycle = 0;

            for (int i = 0; i < NumberOfRepeats; i++)
            {
                CurrentCycle++;
                CoreCommand.Run();

                ProcessAbortAndPauseEvents();

                if (CurrentState == State.Failed || CurrentState == State.Aborted ||
                    CoreCommand.CurrentState == State.Failed || CoreCommand.CurrentState == State.Aborted)
                    break;
            }
        }
    }
}

File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/CyclicCommand.java
package org.extensiblecommands;

/**
 * Implements command to be repeated a specified number of times
 */
public class CyclicCommand extends DecoratorCommand {
    /**
     * How many times to repeat the core command
     */
    private final int numberOfRepeats;

    /**
     * Current cycle number
     */
    private int currentCycle;

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param numberOfRepeats   How many times to repeat the core command
     */
    public CyclicCommand(Command coreCommand, int numberOfRepeats) {
        this(coreCommand, numberOfRepeats, "Cyclic");
    }

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param numberOfRepeats   How many times to repeat the core command
     * @param name              Command name
     */
    public CyclicCommand(Command coreCommand, int numberOfRepeats, String name) {
        super(coreCommand, name);
        this.numberOfRepeats = numberOfRepeats;
    }

    /**
     * @return      How many times to repeat the core command
     */
    public final int getNumberOfRepeats() {
        return numberOfRepeats;
    }

    /**
     * @return      Current cycle number
     */
    public final int getCurrentCycle() {
        return currentCycle;
    }

    @Override
    protected void execute() throws Exception {
        currentCycle = 0;

        for (int i = 0; i < numberOfRepeats; i++) {
            currentCycle++;
            coreCommand.run();

            processAbortAndPauseEvents();

            if (getState() == State.Failed || getState() == State.Aborted ||
                    coreCommand.getState() == State.Failed || coreCommand.getState() == State.Aborted)
                break;
        }
    }
}
